### PR TITLE
adding reindexing capabilities to the backup tool

### PR DESCRIPTION
**Pull request by swsachith**
FIX for BAM-2142: Improve the backup tool to support re-indexing of data

### CODE REVIEW

File: components/analytics-tools/org.wso2.carbon.analytics.tools.backup/src/main/java/org/wso2/carbon/analytics/tools/backup/AnalyticsDataBackupTool.java [hunk 1]
```diff
@@ -119,20 +124,35 @@ public static void main(String[] args) throws Exception {
...
+                backupRecordStore(service, tenantId, baseDir, timeFrom, timeTo, specificTables);
             } else if (line.hasOption("backupFileSystem")) {
                 backupFileSystem(analyticsFileSystem, tenantId, baseDir);
-            } else if (line.hasOption("restoreAnalytics")) {
-                restoreAnalytics(service, tenantId, baseDir, timeFrom, timeTo, specificTables);
+            } else if (line.hasOption("restoreRecordStore")) {
+                restoreRecordStore(service, tenantId, baseDir, timeFrom, timeTo, specificTables);
             } else if (line.hasOption("restoreFileSystem")) {
                 restoreFileSystem(analyticsFileSystem, baseDir);
+            }else if (line.hasOption("reindexEvents")) {
```
Comment (lafernando) [inline components/analytics-tools/org.wso2.carbon.analytics.tools.backup/src/main/java/org/wso2/carbon/analytics/tools/backup/AnalyticsDataBackupTool.java]: Fix code formatting (indentation).

Comment (swsachith) [inline components/analytics-tools/org.wso2.carbon.analytics.tools.backup/src/main/java/org/wso2/carbon/analytics/tools/backup/AnalyticsDataBackupTool.java]: fixed.

File: components/analytics-tools/org.wso2.carbon.analytics.tools.backup/src/main/java/org/wso2/carbon/analytics/tools/backup/AnalyticsDataBackupTool.java [hunk 2]
```diff
@@ -433,21 +459,23 @@ private static void restoreFileStructure(AnalyticsFileSystem analyticsFileSystem
...
+    private static byte[] readFile(File file) throws IOException {
+
+        byte[] buffer = new byte[READ_BUFFER_SIZE];
+        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+        try (InputStream inputStream = new FileInputStream(file)) {
+            int nRead;
+            while ((nRead = inputStream.read(buffer, 0, (int) file.length())) != -1) {
+                byteArrayOutputStream.write(buffer, 0, nRead);
             }
+        } catch (FileNotFoundException e) {
```
Comment (madhawa-gunasekara) [inline components/analytics-tools/org.wso2.carbon.analytics.tools.backup/src/main/java/org/wso2/carbon/analytics/tools/backup/AnalyticsDataBackupTool.java]: shall we move the return statement in to the try block and close the ByteArrayOutputStream in finally block

File: components/analytics-tools/org.wso2.carbon.analytics.tools.backup/src/main/java/org/wso2/carbon/analytics/tools/backup/AnalyticsDataBackupTool.java [hunk 3]
```diff
@@ -53,17 +54,20 @@
...
     private static boolean forceIndexing = false;
     private static final int READ_BUFFER_SIZE = 10;
+    private static final int RECORD_INDEX_CHUNK_SIZE = 1000;
 
     @SuppressWarnings("static-access")
     public static void main(String[] args) throws Exception {
         String timePattern = "yy-mm-dd hh:mm:ss";
         Options options = new Options();
-        options.addOption(new Option("backupAnalytics", false, "backup analytics data"));
+        options.addOption(new Option("backupRecordStore", false, "backup analytics data"));
```
Comment (sinthuja) [inline components/analytics-tools/org.wso2.carbon.analytics.tools.backup/src/main/java/org/wso2/carbon/analytics/tools/backup/AnalyticsDataBackupTool.java]: Make constant for these parameters and re-use it throughout the code.

Comment (swsachith) [inline components/analytics-tools/org.wso2.carbon.analytics.tools.backup/src/main/java/org/wso2/carbon/analytics/tools/backup/AnalyticsDataBackupTool.java]: This has already been fixed in a later PR

**Pull request merged**